File: aoc2025-zig/src/days/day01.zig
const std = @import("std");
const common = @import("../common.zig");
const parse = @import("../parse.zig");

const Instruction = struct { direction: i2, distance: i32 };

pub fn getInstructions(input: [][]const u8, allocator: std.mem.Allocator) ![]Instruction {
    var instructions: std.ArrayList(Instruction) = .empty;
    for (input) |line| {
        var direction: i2 = 1;
        if (line[0] == 'L') {
            direction = -1;
        }
        const distance = std.fmt.parseInt(i32, line[1..], 10) catch unreachable;
        try instructions.append(allocator, Instruction{ .direction = direction, .distance = distance });
    }

    return instructions.toOwnedSlice(allocator);
}

pub fn rotate(instruction: Instruction, current: i32) i32 {
    var res: i32 = current;
    if (instruction.distance >= current and instruction.direction == -1) {
        const difference: i32 = instruction.distance - current;
        res = 100 - @mod(difference, 100);
    } else if (current + instruction.distance >= 100 and instruction.direction == 1) {
        const sum = current + instruction.distance;
        res = @mod(sum - 100, 100);
    } else {
        res = current + (instruction.distance * instruction.direction);
    }

    if (res == -100 or res == 100) {
        res = 0;
    }

    return res;
}

pub fn rotate2(instruction: Instruction, current: i32) struct { dial_value: i32, zero_crossings: usize } {
    var zero_crossings: usize = 0;
    var res: i32 = current;

    if (instruction.direction == -1) {
        if (instruction.distance >= current) {
            if (current != 0 and @mod(instruction.distance, 100) >= current) {
                zero_crossings += 1;
            }
            zero_crossings += @intCast(@divFloor(instruction.distance, 100));
            res = 100 - (@mod(instruction.distance, 100) - current);
        } else {
            res = @mod(current - instruction.distance, 100);
        }
    } else {
        const amount_needed = 100 - current;
        if (instruction.distance >= amount_needed) {
            if (current != 0 and @mod(instruction.distance, 100) >= amount_needed) {
                zero_crossings += 1;
            }
            zero_crossings += @intCast(@divFloor(instruction.distance, 100));
            res = @mod(instruction.distance + current, 100);
        } else {
            res = current + @mod(instruction.distance, 100);
        }
    }

    res = @mod(res, 100);
    return .{ .dial_value = res, .zero_crossings = zero_crossings };
}

const START_VALUE = 50;

pub fn part1(input: [][]const u8, allocator: std.mem.Allocator) ![]const u8 {
    const instructions = try getInstructions(input, allocator);

    var res: i32 = START_VALUE;
    var zero_hits: usize = 0;

    for (instructions) |instruction| {
        res = rotate(instruction, res);
        if (res == 0) {
            zero_hits += 1;
        }
    }

    return std.fmt.allocPrint(allocator, "{d}", .{zero_hits});
}

pub fn part2(input: [][]const u8, allocator: std.mem.Allocator) ![]const u8 {
    const instructions = try getInstructions(input, allocator);

    var dial: i32 = START_VALUE;
    var zero_crossings: usize = 0;

    for (instructions) |instruction| {
        const result = rotate2(instruction, dial);
        dial = result.dial_value;
        zero_crossings += result.zero_crossings;
    }
    return std.fmt.allocPrint(allocator, "{d}", .{zero_crossings});
}

pub fn run(io: std.Io, allocator: std.mem.Allocator) !void {
    const input = try parse.lines(io, allocator, 1, false);

    const t1s = std.Io.Timestamp.now(io, .awake);
    const p1 = try part1(input, allocator);
    const t1e = std.Io.Timestamp.now(io, .awake);

    const t2s = std.Io.Timestamp.now(io, .awake);
    const p2 = try part2(input, allocator);
    const t2e = std.Io.Timestamp.now(io, .awake);

    common.runDay(io, 1, p1, p2, t1s, t1e, t2s, t2e);
}

File: aoc2025-go/days/day01/day01.go
package day01

import (
	"fmt"
	"strconv"
	"time"

	"aoc2025-go/common"
	"aoc2025-go/parse"
)

const DIAL_START = 50

type Rotation struct {
	direction int
	rotations int
}

func get_directions(input []string) []Rotation {
	var directions []Rotation
	for _, line := range input {
		direction := 1
		if line[0] == 'L' {
			direction = -1
		}
		rotations, _ := strconv.Atoi(line[1:])
		directions = append(directions, Rotation{direction, rotations})
	}
	return directions

}

func rotate_dial(r Rotation, current int) int {
	res := current
	if r.rotations >= current && r.direction == -1 {
		difference := r.rotations - current
		res = 100 - (difference % 100)
	} else if current+r.rotations >= 100 && r.direction == 1 {
		sum := current + r.rotations
		res = (sum - 100) % 100
	} else {
		res = current + (r.rotations * r.direction)
	}
	if res == -100 || res == 100 {
		res = 0
	}

	return res
}

func rotate_dial2(r Rotation, current int) (int, int) {
	zero_crossings := 0
	var res int
	if r.direction == -1 {
		if r.rotations >= current {
			if current != 0 && r.rotations%100 >= current {
				zero_crossings += 1
			}
			zero_crossings += r.rotations / 100
			res = 100 - ((r.rotations % 100) - current)
		} else {
			res = (current - r.rotations) % 100
		}
	} else {
		amount_needed := 100 - current
		if r.rotations >= amount_needed {
			if current != 0 && r.rotations%100 >= amount_needed {
				zero_crossings += 1
			}
			zero_crossings += r.rotations / 100
			res = (r.rotations + current) % 100
		} else {
			res = r.rotations + current%100
		}

	}
	res = res % 100

	return res, zero_crossings
}

func Part1(input []string) string {
	_ = input
	directions := get_directions(input)
	zero_hits := 0
	current := DIAL_START
	for _, d := range directions {
		current = rotate_dial(d, current)
		if current == 0 {
			zero_hits++
		}
	}

	return fmt.Sprintf("%d", zero_hits)
}

func Part2(input []string) string {
	directions := get_directions(input)
	zero_hits := 0
	current := DIAL_START
	for _, d := range directions {
		res, zero_crossings := rotate_dial2(d, current)
		zero_hits += zero_crossings
		current = res
	}
	return fmt.Sprintf("%d", zero_hits)
}

func Run() {
	input := parse.Lines(1, false)

	t1s := time.Now()
	p1 := Part1(input)
	t1 := time.Since(t1s)

	t2s := time.Now()
	p2 := Part2(input)
	t2 := time.Since(t2s)

	common.RunDay(1, p1, p2, t1, t2)
}
